network/default.go

```diff
@@ -0,0 +1,389 @@
+	for i, record := range records {
+		nodes[i] = record.Address
+	}
+
+	return nodes, nil
+}
+
+func (n *network) resolve() {
+	resolve := time.NewTicker(ResolveTime)
+	defer resolve.Stop()
+
+	for {
+		select {
+		case <-n.closed:
+			return
+		case <-resolve.C:
+			nodes, err := n.resolveNodes()
+			if err != nil {
+				log.Debugf("Network failed to resolve nodes: %v", err)
+				continue
+			}
+			// initialize the tunnel
+			n.Tunnel.Init(
+				tunnel.Nodes(nodes...),
+			)
+		}
+	}
+}
+
+func (n *network) handleConn(conn tunnel.Conn, msg chan *transport.Message) {
+	for {
+		m := new(transport.Message)
+		if err := conn.Recv(m); err != nil {
+			// TODO: should we bail here?
+			log.Debugf("Network tunnel advert receive error: %v", err)
+			return
+		}
+
+		select {
+		case msg <- m:
+		case <-n.closed:
+			return
+		}
+	}
+}
+
+func (n *network) process(l tunnel.Listener) {
+	// receive control message queue
+	recv := make(chan *transport.Message, 128)
+
+	// accept a connection
+	conn, err := l.Accept()
+	if err != nil {
+		// TODO: handle this
+		log.Debugf("Network tunnel accept error: %v", err)
+		return
+	}
+
+	go n.handleConn(conn, recv)
+
+	for {
+		select {
+		case m := <-recv:
+			// switch on type of message and take action
+			switch m.Header["Micro-Method"] {
+			case "advert":
+				pbAdvert := &pb.Advert{}
+				if err := proto.Unmarshal(m.Body, pbAdvert); err != nil {
+					continue
+				}
+
+				var events []*router.Event
+				for _, event := range pbAdvert.Events {
+					route := router.Route{
+						Service: event.Route.Service,
+						Address: event.Route.Address,
+						Gateway: event.Route.Gateway,
+						Network: event.Route.Network,
+						Link:    event.Route.Link,
+						Metric:  int(event.Route.Metric),
+					}
+					e := &router.Event{
+						Type:      router.EventType(event.Type),
+						Timestamp: time.Unix(0, pbAdvert.Timestamp),
+						Route:     route,
+					}
+					events = append(events, e)
+				}
+				advert := &router.Advert{
+					Id:        pbAdvert.Id,
+					Type:      router.AdvertType(pbAdvert.Type),
+					Timestamp: time.Unix(0, pbAdvert.Timestamp),
+					TTL:       time.Duration(pbAdvert.Ttl),
+					Events:    events,
+				}
+
+				if err := n.Router.Process(advert); err != nil {
+					log.Debugf("Network failed to process advert %s: %v", advert.Id, err)
+					continue
+				}
+			}
+		case <-n.closed:
+			return
+		}
+	}
+}
+
+// advertise advertises routes to the network
+func (n *network) advertise(client transport.Client, advertChan <-chan *router.Advert) {
+	for {
+		select {
+		// process local adverts and randomly fire them at other nodes
+		case advert := <-advertChan:
+			// create a proto advert
+			var events []*pb.Event
+			for _, event := range advert.Events {
+				route := &pb.Route{
+					Service: event.Route.Service,
+					Address: event.Route.Address,
+					Gateway: event.Route.Gateway,
+					Network: event.Route.Network,
+					Link:    event.Route.Link,
```

Comment:
Should we set the link here to "network" or on receiving?

Reply:
Not sure. On send we set it to "network" saying, hey this is a "remote" route. On receive you should just process the routes as you received them, no?

Reply:
Network field should not be modified. Link field should be set to "Network". This may be why it's not working.

Reply:
Yeah I got Network and Link mixed up -- my brain short-circuited the string "network" with `Route.Network` lmao. Nice one!